cpp/python/include/ucxx/python/python_future_task.h

```diff
@@ -0,0 +1,213 @@
+      setPythonException(PyExc_ValueError, e.what());
+    } catch (const std::invalid_argument& e) {
+      setPythonException(PyExc_ValueError, e.what());
+    } catch (const std::ios_base::failure& e) {
+      setPythonException(PyExc_IOError, e.what());
+    } catch (const std::out_of_range& e) {
+      setPythonException(PyExc_IndexError, e.what());
+    } catch (const std::overflow_error& e) {
+      setPythonException(PyExc_OverflowError, e.what());
+    } catch (const std::range_error& e) {
+      setPythonException(PyExc_ArithmeticError, e.what());
+    } catch (const std::underflow_error& e) {
+      setPythonException(PyExc_ArithmeticError, e.what());
+    } catch (const std::exception& e) {
+      setPythonException(PyExc_RuntimeError, e.what());
+    } catch (...) {
+      setPythonException(PyExc_RuntimeError, "Unknown exception");
+    }
+  }
+
+ public:
+  /**
+   * @brief Construct a Python future backed by C++ `std::packaged_task`.
+   *
+   * Construct a future object that receives a user-defined C++ `std::packaged_task` which
+   * runs asynchronously using an internal `std::async` that ultimately notifies a Python
+   * future that can be awaited in Python code.
+   *
+   * @param[in] task the user-defined C++ task.
+   * @param[in] pythonConvert C-Python function to convert a C object into a `PyObject*`
+   *                          representing the result of the task.
+   * @param[in] asyncioEventLoop pointer to a valid Python object containing the event loop
+   *                             that the application is using, to which the Python future
+   *                             will belong to.
+   * @param[in] launchPolicy launch policy for the async C++ task.
+   */
+  explicit PythonFutureTask(std::packaged_task<ReturnType(TaskArgs...)> task,
+                            std::function<PyObject*(ReturnType)> pythonConvert,
+                            PyObject* asyncioEventLoop,
+                            std::launch launchPolicy = std::launch::async)
+    : _task{std::move(task)},
+      _pythonConvert(pythonConvert),
+      _asyncioEventLoop(asyncioEventLoop),
+      _handle{ucxx::python::create_python_future_with_event_loop(asyncioEventLoop)}
+  {
+    _future = std::async(launchPolicy, [this]() {
+      std::future<ReturnType> result = this->_task.get_future();
+      this->_task();
+      try {
+        const ReturnType r = result.get();
+        this->setResult(r);
+        return r;
+      } catch (std::exception& e) {
+        this->setException(e);
+      }
+    });
+  }
+  PythonFutureTask(const PythonFutureTask&)            = delete;
+  PythonFutureTask& operator=(PythonFutureTask const&) = delete;
+  PythonFutureTask(PythonFutureTask&& o)               = delete;
+  PythonFutureTask& operator=(PythonFutureTask&& o)    = delete;
+
+  /**
+   * @brief Python future destructor.
+   *
+   * Decrement the reference count of the underlying Python future.
+   */
+  ~PythonFutureTask() { Py_XDECREF(_handle); }
```

Comment:
I guess it is safe as long as the current thread is able to take the GIL, i.e. the "main" thread releases the GIL at some point. However, that doesn't mean it couldn't cause deadlocks depending on how the application makes use of that. Definitely a Python cleanup routine could make things safer but I can't think of a cleanup routine that would not require some complex Python mechanism, meaning it can't be implemented here alone, perhaps you already have something in mind? The only alternative I can think of, which is definitely not a good one, would be to implement a public `release()` method that the user is supposed to call where it is known that taking the GIL is safe and then change the destructor to print a warning/error if `release()` wasn't called.

Reply:
Sorry, we already have a `release()` which has a different meaning (release the handle to the user). Replace `release()` in my comment above with `free()`.

Reply:
Do we need `_handle` after `setResult()` or `setPythonException()` has been called? If not, maybe let them release the handle?

Reply:
The `_handle` is the `asyncio.Future` object, so if the user never got a reference to it via `getHandle()`/`release()` before the C++ future completed and notified the Python future then the user will never be able to get it later if we release it in `setResult()`/`setPythonException()`. Therefore, I don't think this would work.

Reply:
I see, that won't work :)
The only alternative to implementing a`free()` method, is to save the `_handle` in a global struct that we can clean up periodically.

E.g. if `PyGILState_Check() == 0`, we add `_handle` to a global struct that is checked in some of our API calls that grabs the GIL.

Reply:
Agreed, that's probably our best option. It will take me until next week or the week after to get back to this, but I'll try to make it work.

Reply:
I have now [introduced a new singleton class `PythonFutureTaskCollector`](https://github.com/rapidsai/ucxx/pull/46/commits/72eb53938496b34db64f71e9d6ca151e339b6996) which allows pushing `PyObject*` when `~PythonFutureTask()` runs instead of it requiring the GIL to decref those. The GIL and decref now happens when `PythonFutureTaskCollector::collect()` gets called, but that needs to be done by the application itself, which [currently runs it during `~ApplicationThread`](https://github.com/rapidsai/ucxx/pull/46/commits/6e6314033b01e4efed1d395b5c0c5f0e1c1276e0).

This is not an optimal solution, but I think it's probably the only one that guarantees objects are actually decrefed at some point, the alternative may be to register another `atexit` handler. A periodic call of that method could also improve things a bit during runtime but is not enough to ensure everything gets cleaned up before exiting, as we may always miss the last few destroyed `PythonFutureTask` if the periodic call happens to be called one last time before some `~PythonFutureTask`. Perhaps I'm overlooking some obvious better and guaranteed way to ensure we don't leave any `PyObject*` behind, if you can think of any please let me know @madsbk @wence- .

Reply:
This kind of deferred collection with stop-the-world cleanup points is the way I (with others) solve morally the same problem in a distributed setting (where two processes have to agree to destruct objects collectively). This is kind of similar because we're needing to do collection at a point when we can guarantee synchronisation of some resource (in this case the gil + any other locks that might be required).

So I think this is fine. You might imagine that you'd want to also call this "collect" periodically during submission of tasks if that happens on the main thread that will hold the gil.

Reply:
> So I think this is fine. You might imagine that you'd want to also call this "collect" periodically during submission of tasks if that happens on the main thread that will hold the gil.

I agree with that idea, what I'm not sure is who should be responsible for that. I can think of a few different approaches:

1. Within the `PythonFutureTask`, for example every time a new one is constructed;
2. As part of the `ApplicationThread`, perhaps during `progressUntilSync()`;
3. Directly from Python with something similar to [`tornado.ioloop.PeriodicCallback`](https://www.tornadoweb.org/en/stable/ioloop.html#tornado.ioloop.PeriodicCallback);
4. Directly from Python with looping forever submitting a new async task for every iteration, similar to [`PollingMode`](https://github.com/rapidsai/ucxx/blob/branch-0.36/python/ucxx/_lib_async/continuous_ucx_progress.py#L64-L72).

I think we should be doing one or more of those, but it's difficult to decide the best approach. Furthermore, `ApplicationThread` is meant to be an example only, how to actually do it probably depends on the actual application needs so we don't need to cover every possible alternative. We will need to find an appropriate solution to this for UCXX though, but that will happen in a follow-up PR.